src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs

```diff
@@ -50,25 +52,31 @@ public AllLowerCamelCaseMatcher(string candidate, bool includeMatchedSpans, Stri
                 var patternIndex = 0;
                 var candidateHumpIndex = 0;
 
-                var (bestWeight, localMatchedSpans) = TryMatch(
+                var (bestWeight, localMatchedSpansInReverse) = TryMatch(
                    patternIndex, candidateHumpIndex, contiguous: null);
 
-                matchedSpans = localMatchedSpans;
+                matchedSpans = _includeMatchedSpans && localMatchedSpansInReverse != null
+                    ? new NormalizedTextSpanCollection(localMatchedSpansInReverse).ToImmutableArray()
+                    : ImmutableArray<TextSpan>.Empty;
+
+                localMatchedSpansInReverse?.Free();
                 return bestWeight;
             }
 
-            private (int? bestWeight, List<TextSpan> matchedSpans) TryMatch(
+            private (int? bestWeight, ArrayBuilder<TextSpan> matchedSpansInReverse) TryMatch(
                 int patternIndex, int candidateHumpIndex, bool? contiguous)
             {
                 if (patternIndex == _patternText.Length)
                 {
                     // We hit the end.  So we were able to match against this candidate.
-                    return (bestWeight: contiguous == false ? 0 : CamelCaseContiguousBonus,
-                            matchedSpans: _includeMatchedSpans ? new List<TextSpan>() : null);
+                    var weight = contiguous == false ? 0 : CamelCaseContiguousBonus;
+                    var matchedSpansInReverse = _includeMatchedSpans ? ArrayBuilder<TextSpan>.GetInstance() : null;
+
+                    return (weight, matchedSpansInReverse);
                 }
 
                 var bestWeight = default(int?);
-                var bestMatchedSpans = default(List<TextSpan>);
+                var bestMatchedSpansInReverse = default(ArrayBuilder<TextSpan>);
```

Comment:
😦 Not a fan of using `default(T)` for cases where `T` is known to be a reference type. It confuses the syntax with cases like `default(ImmutableArray<int>)` which produces a non-null value with "interesting" semantics. The above line could be improved by writing it as the following, even though it no longer uses `var`.

```csharp
ArrayBuilder<TextSpan> bestMatchedSpansInReverse = null;
```